File: Circles.Profiles.Sdk/DefaultNamespaceReader.cs
using Circles.Profiles.Interfaces;
using Circles.Profiles.Models;
using Circles.Profiles.Models.Core;
using Circles.Profiles.Sdk.Utils;
using Nethereum.Hex.HexConvertors.Extensions;

namespace Circles.Profiles.Sdk;

/// <summary>
/// Streams a namespace **with on‑the‑fly signature verification**.
/// Links failing verification are silently skipped.
/// </summary>
public sealed class DefaultNamespaceReader : INamespaceReader
{
    private readonly IIpfsStore _ipfs;
    private readonly ISignatureVerifier _verifier;
    private readonly string? _headCid;

    public DefaultNamespaceReader(string? headCid,
        IIpfsStore ipfs,
        ISignatureVerifier verifier)
    {
        _headCid = headCid;
        _ipfs = ipfs;
        _verifier = verifier;
    }

    public async Task<CustomDataLink?> GetLatestAsync(
        string logicalName, CancellationToken ct = default)
    {
        await foreach (var l in StreamAsync(0, ct))
        {
            if (l.Name.Equals(logicalName, StringComparison.OrdinalIgnoreCase))
            {
                return l;
            }
        }

        return null;
    }

    public async IAsyncEnumerable<CustomDataLink> StreamAsync(
        long newerThanUnixTs = 0,
        [System.Runtime.CompilerServices.EnumeratorCancellation]
        CancellationToken ct = default)
    {
        for (string? cur = _headCid; cur is not null;)
        {
            var chunk = await Helpers.LoadChunk(cur, _ipfs, ct);

            foreach (var l in chunk.Links.OrderByDescending(l => l.SignedAt))
            {
                if (l.SignedAt <= newerThanUnixTs)
                {
                    continue;
                }

                if (!await Verify(l, ct))
                {
                    continue;
                }

                yield return l;
            }

            cur = chunk.Prev;
        }
    }

    private async Task<bool> Verify(CustomDataLink l, CancellationToken ct)
    {
        bool seenBefore = NonceRegistrySingleton.Instance.SeenBefore(l.Nonce);
        if (seenBefore)
        {
            return false; // replay → drop
        }

        // Compute canonical payload bytes once for both paths
        byte[] payloadBytes = CanonicalJson.CanonicaliseWithoutSignature(l);
        byte[] payloadHash = Sha3.Keccak256Bytes(payloadBytes);
        byte[] signature = l.Signature.HexToByteArray();

        // Primary path: works for EOAs and many contracts
        bool primaryOk = await _verifier.VerifyAsync(payloadHash, l.SignerAddress, signature, ct);
        if (primaryOk)
        {
            return true;
        }

        // Secondary path for Safe messages: prefer a verifier that supports the "bytes" variant
        bool hasSafeVerifier = _verifier is ISafeBytesVerifier;
        if (hasSafeVerifier)
        {
            var safe = (ISafeBytesVerifier)_verifier;
            bool bytesOk = await safe.Verify1271WithBytesAsync(payloadBytes, l.SignerAddress, signature, ct);
            if (bytesOk)
            {
                return true;
            }
        }

        return false;
    }
}

File: js/src/DefaultNamespaceReader.ts
import { CustomDataLink } from './interfaces/CustomDataLink';
import { IIpfsStore } from './interfaces/IIpfsStore';
import { INamespaceReader } from './interfaces/INamespaceReader';
import { ISignatureVerifier } from './interfaces/ISignatureVerifier';
import { CanonicalJson } from './CanonicalJson';
import { Sha3 } from './Sha3';
import { Helpers } from './utils/Helpers';

/**
 * Streams a namespace with on-the-fly signature verification.
 * Links failing verification are silently skipped.
 */
export class DefaultNamespaceReader implements INamespaceReader {
  private readonly ipfs: IIpfsStore;
  private readonly verifier: ISignatureVerifier;
  private readonly headCid: string | null;
  
  /**
   * Creates a new DefaultNamespaceReader.
   * 
   * @param headCid The CID of the head chunk
   * @param ipfs The IPFS store
   * @param verifier The signature verifier
   */
  constructor(
    headCid: string | null,
    ipfs: IIpfsStore,
    verifier: ISignatureVerifier
  ) {
    this.headCid = headCid;
    this.ipfs = ipfs;
    this.verifier = verifier;
  }
  
  /**
   * Gets the latest link with the given logical name.
   * 
   * @param logicalName The logical name to look for
   * @returns The latest link, or null if not found
   */
  public async getLatestAsync(
    logicalName: string
  ): Promise<CustomDataLink | null> {
    for await (const link of this.streamAsync(0)) {
      if (link.name.toLowerCase() === logicalName.toLowerCase()) {
        return link;
      }
    }
    
    return null;
  }
  
  /**
   * Streams all links newer than the given timestamp.
   * 
   * @param newerThanUnixTs Optional timestamp to filter by
   * @returns Async iterable of links
   */
  public async* streamAsync(
    newerThanUnixTs: number = 0
  ): AsyncIterable<CustomDataLink> {
    let currentCid = this.headCid;
    
    while (currentCid) {
      const chunk = await Helpers.loadChunk(currentCid, this.ipfs);
      
      // Process links in descending order of timestamp
      // For links with the same timestamp, preserve original order
      const sortedLinks = [...chunk.links].sort((a, b) => {
        const timestampDiff = b.signedAt - a.signedAt;
        return timestampDiff !== 0 ? timestampDiff : 0; // Keep original order if timestamps match
      });
      
      for (const link of sortedLinks) {
        if (link.signedAt <= newerThanUnixTs) continue;
        
        if (await this.verify(link)) {
          yield link;
        }
      }
      
      currentCid = chunk.prev;
    }
  }
  
  /**
   * Verifies a link's signature.
   * 
   * @param link The link to verify
   * @returns True if the signature is valid, false otherwise
   */
  private async verify(link: CustomDataLink): Promise<boolean> {
    if (!link.signature || !link.signerAddress) {
      return false;
    }
    
    const hash = Sha3.keccak256Bytes(
      CanonicalJson.canonicaliseWithoutSignature(link)
    );
    
    const signature = Buffer.from(link.signature.startsWith('0x')
      ? link.signature.slice(2)
      : link.signature, 'hex');
    
    return await this.verifier.verifyAsync(
      hash,
      link.signerAddress,
      signature
    );
  }
}
